zerver/lib/message.py

```diff
@@ -407,6 +408,82 @@ def is_subscribed_helper() -> bool:
+        """Returns the IDs of users who actually received the messages."""
+        usermessages = UserMessage.objects.filter(message_id__in=message_ids)
+        if exclude_long_term_idle_users:
+            usermessages = usermessages.exclude(user_profile__long_term_idle=True)
+        return set(usermessages.values_list("user_profile_id", flat=True))
+
+    sample_message = messages[0]
+    if not sample_message.is_stream_message():
+        # For DM, event is sent to users who actually received the message.
+        return get_user_ids_having_usermessage_row_for_messages(message_ids)
+
+    channel_id = sample_message.recipient.type_id
+    if channel is None:
+        channel = Stream.objects.get(id=channel_id)
+
+    subscriptions = get_active_subscriptions_for_stream_id(
+        channel_id, include_deactivated_users=False
+    )
+    if exclude_long_term_idle_users:
+        subscriptions = subscriptions.exclude(user_profile__long_term_idle=True)
+    subscriber_ids = set(subscriptions.values_list("user_profile_id", flat=True))
+
+    if channel.is_history_public_to_subscribers():
+        if channel.is_public():
+            # For public channel with public history, event is sent to:
+            # 1. Subscribers
+            # 2. Unsubscribed users having usermessage row & channel access.
+            #    * Users who never subscribed but starred or reacted on messages
+            #      (usermessages with historical flag exists for such cases).
+            #    * Users who were initially subscribed and later unsubscribed
+            #      (usermessages exist for messages they received while subscribed).
+            # We don't send it to all unsubscribed users as a performance tradeoff.
+            usermessage_rows = (
+                UserMessage.objects.filter(message_id__in=message_ids)
+                .select_related("user_profile__realm")
+                .only(
+                    "user_profile__long_term_idle",
+                    "user_profile__role",
+                    "user_profile__realm__string_id",
+                )
+            )
+            if exclude_long_term_idle_users:
+                usermessage_rows = usermessage_rows.exclude(user_profile__long_term_idle=True)
+            user_ids_with_usermessage_row_and_channel_access = {
+                usermessage.user_profile.id
+                for usermessage in usermessage_rows
+                if usermessage.user_profile.can_access_public_streams()
+            }
+            return user_ids_with_usermessage_row_and_channel_access | subscriber_ids
+
+        # For private channel with NOT protected history,
+        # all the subscribers are notified.
+        return subscriber_ids
+
+    # For protected history, only users who are subscribed and
+    # received the original message are notified.
+    user_ids_with_usermessage_row = get_user_ids_having_usermessage_row_for_messages(message_ids)
```

Comment:
I reordered this function to make better use of the early return pattern, and simplify the queries. Also set the default to `True` as discussed previously, and dropped the explicit from the one caller where you'd passed that.